// ng2-3d-editor/src/controllers/editorControls.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var THREE = require("three");
var EditorControls = (function (_super) {
    __extends(EditorControls, _super);
    function EditorControls(container, camera) {
        var _this = _super.call(this) || this;
        _this.panSpeed = 0.001;
        _this.zoomSpeed = 0.001;
        _this.rotationSpeed = 0.005;
        _this.pointer = new THREE.Vector2();
        _this.pointerOld = new THREE.Vector2();
        _this.STATE = { NONE: -1, ROTATE: 0, ZOOM: 1, PAN: 2 };
        _this.center = new THREE.Vector3();
        _this.vector = new THREE.Vector3();
        _this.spherical = new THREE.Spherical();
        _this.changeEvent = { type: 'change' };
        _this.normalMatrix = new THREE.Matrix3();
        _this.state = _this.STATE.NONE;
        _this.center = new THREE.Vector3();
        _this.camera = camera;
        _this.container = container;
        _this.container.addEventListener('contextmenu', _this.contextmenu.bind(_this), false);
        _this.container.addEventListener('mousedown', _this.onMouseDown.bind(_this), false);
        _this.container.addEventListener('wheel', _this.onMouseWheel.bind(_this), false);
        return _this;
    }
    EditorControls.prototype.onMouseDown = function (event) {
        if (event.button === 0) {
            this.state = this.STATE.ROTATE;
        }
        else if (event.button === 1) {
            this.state = this.STATE.ZOOM;
        }
        else if (event.button === 2) {
            this.state = this.STATE.PAN;
        }
        this.pointerOld.set(event.clientX, event.clientY);
        this.container.addEventListener('mousemove', this.onMouseMove.bind(this), false);
        this.container.addEventListener('mouseup', this.onMouseUp.bind(this), false);
        this.container.addEventListener('mouseout', this.onMouseUp.bind(this), false);
        this.container.addEventListener('dblclick', this.onMouseUp.bind(this), false);
    };
    EditorControls.prototype.onMouseMove = function (event) {
        this.pointer.set(event.clientX, event.clientY);
        var movementX = this.pointer.x - this.pointerOld.x;
        var movementY = this.pointer.y - this.pointerOld.y;
        if (this.state === this.STATE.ROTATE) {
            this.rotate(new THREE.Vector3(-movementX * this.rotationSpeed, -movementY * this.rotationSpeed, 0));
        }
        else if (this.state === this.STATE.ZOOM) {
            this.zoom(new THREE.Vector3(0, 0, movementY));
        }
        else if (this.state === this.STATE.PAN) {
            this.pan(new THREE.Vector3(-movementX, movementY, 0));
        }
        this.pointerOld.set(event.clientX, event.clientY);
    };
    EditorControls.prototype.focus = function (target) {
        var box = new THREE.Box3().setFromObject(target);
        this.camera.lookAt(this.center.copy(box.getCenter()));
        this.dispatchEvent(this.changeEvent);
    };
    EditorControls.prototype.dispatchEvent = function (event) {
        _super.prototype.dispatchEvent.call(this, event);
    };
    EditorControls.prototype.pan = function (delta) {
        var distance = this.camera.position.distanceTo(this.center);
        delta.multiplyScalar(distance * this.panSpeed);
        delta.applyMatrix3(this.normalMatrix.getNormalMatrix(this.camera.matrix));
        this.camera.position.add(delta);
        this.center.add(delta);
        this.dispatchEvent(this.changeEvent);
    };
    EditorControls.prototype.zoom = function (delta) {
        var distance = this.camera.position.distanceTo(this.center);
        delta.multiplyScalar(distance * this.zoomSpeed);
        if (delta.length() > distance) {
            return;
        }
        delta.applyMatrix3(this.normalMatrix.getNormalMatrix(this.camera.matrix));
        this.camera.position.add(delta);
        this.dispatchEvent(this.changeEvent);
    };
    EditorControls.prototype.rotate = function (delta) {
        this.vector.copy(this.camera.position).sub(this.center);
        this.spherical.setFromVector3(this.vector);
        this.spherical.theta += delta.x;
        this.spherical.phi += delta.y;
        this.spherical.makeSafe();
        this.vector.setFromSpherical(this.spherical);
        this.camera.position.copy(this.center).add(this.vector);
        this.camera.lookAt(this.center);
        this.dispatchEvent(this.changeEvent);
    };
    EditorControls.prototype.onMouseWheel = function (event) {
        event.preventDefault();
        this.zoom(new THREE.Vector3(0, 0, event.deltaY));
    };
    EditorControls.prototype.onMouseUp = function () {
        this.container.removeEventListener('mousemove', this.onMouseMove.bind(this), false);
        this.container.removeEventListener('mouseup', this.onMouseUp.bind(this), false);
        this.container.removeEventListener('mouseout', this.onMouseUp.bind(this), false);
        this.container.removeEventListener('dblclick', this.onMouseUp.bind(this), false);
        this.state = this.STATE.NONE;
    };
    EditorControls.prototype.contextmenu = function (event) {
        event.preventDefault();
    };
    return EditorControls;
}(THREE.EventDispatcher));
EditorControls = __decorate([
    core_1.Injectable(),
    __metadata("design:paramtypes", [Object, Object])
], EditorControls);
exports.EditorControls = EditorControls;
