// ng2-3d-editor/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var core_1 = require("@angular/core");
var viewer3D_component_1 = require("./src/componets/viewer3D.component");
var common_1 = require("@angular/common");
__export(require("./src/componets/viewer3D.component"));
exports.VIEWER_DIRECTIVES = [
    viewer3D_component_1.Viewer3DComponent
];
var Editor3DModule = Editor3DModule_1 = (function () {
    function Editor3DModule() {
    }
    Editor3DModule.forRoot = function () {
        return {
            ngModule: Editor3DModule_1
        };
    };
    return Editor3DModule;
}());
Editor3DModule = Editor3DModule_1 = __decorate([
    core_1.NgModule({
        imports: [
            common_1.CommonModule
        ],
        declarations: exports.VIEWER_DIRECTIVES.slice(),
        exports: exports.VIEWER_DIRECTIVES.slice()
    })
], Editor3DModule);
exports.Editor3DModule = Editor3DModule;
var Editor3DModule_1;
